core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,346 @@
+package audit
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+
+	"github.com/pkg/errors"
+)
+
+const bufferCapacity = 2048
+const webRequestTimeout = 10
+
+type Data = map[string]any
+
+type AuditLogger interface {
+	services.ServiceCtx
+
+	Audit(eventID EventID, data Data)
+}
+
+type validatedAuditLoggerConfig struct {
+	forwardToUrl   string
+	environment    string
+	jsonWrapperKey string
+	headers        []ServiceHeader
+}
+
+type AuditLoggerConfig struct {
+	ForwardToUrl   *string
+	Environment    *string
+	JsonWrapperKey *string
+	Headers        *string
+}
+
+type HTTPAuditLoggerInterface interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// / Wrap all the items into a struct to easily pass them around. This does not
+// / do any validation on the values given
+func NewAuditLoggerConfig(forwardToUrl string, isDev bool, jsonWrapperKey string, headers string) AuditLoggerConfig {
+	environment := "production"
+	if isDev {
+		environment = "develop"
+	}
+
+	return AuditLoggerConfig{
+		ForwardToUrl:   &forwardToUrl,
+		Environment:    &environment,
+		JsonWrapperKey: &jsonWrapperKey,
+		Headers:        &headers,
+	}
+}
+
+// / Internal function to validate the values sent in to the AuditLogger
+// /
+// / Checks all sorts of things to eliminate nil pointers and invalid endpoints
+// / as well as parsing out the headers into format that's easier to use later.
+func validateConfig(alc AuditLoggerConfig) (validatedAuditLoggerConfig, error) {
```

Comment:
This is functional for the current legacy configuration, but it is unfortunately no longer compatible with the new TOML config.

Reply:
It may be simpler to split the types between toml and legacy if that helps, but I really think that most of these problems would go away if you switched to the usual convention of receiving an interface of individual methods. :shrug:

Reply:
What is incompatible about this function?

Reply:
The new TOML config does validation up front, and per-field via types. Using plain strings and manually validating later breaks this model. The previous version was very close to what we needed.

Reply:
We need to manually validate later or duplicate the validation code because the fields are public

Reply:
I don't understand what you mean. If we enforce as much as possible via strict types in the config struct/interface then the validation code can be limited to `UnmarshalText` implementations and trivially reused to parse from anywhere else if necessary.

Reply:
See `models.URL` for example:
https://github.com/smartcontractkit/chainlink/blob/af5b414cb86677a86f20993ded92fcce63ad11ca/core/store/models/common.go#L512

Reply:
There might still be some duplication for higher level validation, but if so then it's just necessary until we retire legacy config :shrug: 
It's also possible the new config `ValidateConfig()` impl would be reusable, but it has to exist for new config going forward either way.